### ISSUE TITLE
Request for adding City Of Lincoln, UK

### ISSUE THREAD

**Issue by RoseOO** [mampfes/hacs_waste_collection_schedule#1246]
Hi,

Just looking into this but Lincoln UK does have an online service to provide bin collection dates:

[HERE](https://contact.lincoln.gov.uk/AchieveForms/?mode=fill&consentMessage=yes&form_uri=sandbox-publish://AF-Process-503f9daf-4db9-4dd8-876a-6f2029f11196/AF-Stage-a1c0af0f-fec1-4419-80c0-0dd4e1d965c9/definition.json&process=1&process_uri=sandbox-processes://AF-Process-503f9daf-4db9-4dd8-876a-6f2029f11196&process_id=AF-Process-503f9daf-4db9-4dd8-876a-6f2029f11196)

A sample postcode to use is LN5 7SH

The call to get the data is a bit of a CRM mess, probably a lot can be removed. I've included a sample below there is a JSON here is a small portion:

```json
"000235024841": {
                    "display": "10 Sibthorp Street Lincoln Lincolnshire LN5 7SH ",
                    "name": "000235024841",
                    "Street": "Sibthorp Street",
                    "Locality": "",
                    "Postcode": "LN5 7SH",
                    "refuse_schedule": "wy_wed",
                    "refusenextdate": "2023-09-13",
                    "refuse_freq": "weekly",
                    "refuse_day": "Wednesday",
                    "refusenextcount": "6",
                    "recycling_schedule": "w2_wed",
                    "recyclenextdate": "2023-09-20",
                    "recycle_freq": "fortnightly",
                    "recycle_day": "Wednesday",
                    "recyclenextcount": "13",
                    "garden_schedule": "w2_mon",
                    "gardennextdate": "2023-09-18",
                    "garden_freq": "fortnightly",
                    "garden_day": "Monday",
                    "gardennextcount": "11"
                },
```

_________

```bash
curl 'https://contact.lincoln.gov.uk/apibroker/runLookup?id=62aafd258f72c&repeat_against=&noRetry=false&getOnlyTokens=undefined&log_id=&app_name=AF-Renderer::Self' \
  -H 'authority: contact.lincoln.gov.uk' \
  -H 'accept: */*' \
  -H 'accept-language: en,en-GB;q=0.9,fr;q=0.8' \
  -H 'content-type: application/json' \
  -H 'cookie: legalmonster-cookie-consent={%22LoAwoTYsh8cGcMMMTArDuAKt%22:{%22cookieConsentState%22:{%22wasDoNotTrackSet%22:false%2C%22analytics%22:true%2C%22marketing%22:true}%2C%22expires%22:1701441928521%2C%22lastAccessed%22:1685889928521%2C%22version%22:2}}; COOKIE_MESSAGE=yes; legalmonster-user={%22LoAwoTYsh8cGcMMMTArDuAKt%22:{%22userId%22:%22oSrJexbBcWUppEygLiTEzrNN%22%2C%22expires%22:1709680736000%2C%22lastAccessed%22:1694128736000}}; legalmonster-pages-viewed=1; PHPSESSID=bat4bpe49e41sgp15kps52ueid; APP_LANG=en; product=SELF; SimpleSAMLSessionID=64efaf9224648cae85be2e5c2c54bb22; localtime=2023-09-08 00:22:05; AWSALB=e0aUHLsdphAZcLodhHoa/Mn/RJfiYQ09c1ZljcEQle4wdLWzBBSQsgDfytt2dm9YNa5mfxWy3/2FNAWF0CdktDLFcNzu30geoER0apFqoJe05sRAZr3OmFEB/qTK; AWSALBCORS=e0aUHLsdphAZcLodhHoa/Mn/RJfiYQ09c1ZljcEQle4wdLWzBBSQsgDfytt2dm9YNa5mfxWy3/2FNAWF0CdktDLFcNzu30geoER0apFqoJe05sRAZr3OmFEB/qTK' \
  -H 'dnt: 1' \
  -H 'origin: https://contact.lincoln.gov.uk' \
  -H 'referer: https://contact.lincoln.gov.uk/fillform/?iframe_id=fillform-frame-1&db_id=' \
  -H 'sec-ch-ua: "Chromium";v="116", "Not)A;Brand";v="24", "Google Chrome";v="116"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: same-origin' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36' \
  -H 'x-requested-with: XMLHttpRequest' \
  --data-raw '{"stopOnFailure":true,"usePHPIntegrations":true,"stage_id":"AF-Stage-a1c0af0f-fec1-4419-80c0-0dd4e1d965c9","stage_name":"Stage 1","formId":"AF-Form-6356dd50-b92b-48c8-a766-f775353ed997","formValues":{"Section 1":{"postcode":{"name":"postcode","type":"text","id":"AF-Field-3958e8bf-fb8f-4f2e-acd6-fd00c3925cda","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"Street Name or Postcode","value_label":"","hasOther":false,"value":"LN5 7SH","path":"root/address/postcode","valid":true,"totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":false,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"chooseaddress":{"name":"chooseaddress","type":"select","id":"AF-Field-d835497e-5cfd-4410-8bd3-650fb59edb8e","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"Select address","value_label":"","hasOther":false,"value":"","path":"root/address/chooseaddress","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":false,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"chooseaddressdisplay":{"name":"chooseaddressdisplay","type":"text","id":"AF-Field-41e5b47e-89cb-48f5-b2dd-7ae1f72a34eb","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"chooseaddressdisplay","value_label":"","hasOther":false,"value":"","path":"root/address/chooseaddressdisplay","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":false,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"refusenextdate":{"name":"refusenextdate","type":"text","id":"AF-Field-78089a10-1e8e-4db2-b22d-0326c5d6564b","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"refusenextdate","value_label":"","hasOther":false,"value":"","path":"root/address/refusenextdate","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"refuse_freq":{"name":"refuse_freq","type":"text","id":"AF-Field-d840c858-dce5-4113-8091-50968acf361c","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"refuse_freq","value_label":"","hasOther":false,"value":"","path":"root/address/refuse_freq","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"refuse_day":{"name":"refuse_day","type":"text","id":"AF-Field-8fa1643b-490c-44d9-ae5e-bdebac086163","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"refuse_day","value_label":"","hasOther":false,"value":"","path":"root/address/refuse_day","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"refusenextcount":{"name":"refusenextcount","type":"text","id":"AF-Field-a5a1cf92-90bf-4a7b-abd9-1843fab67ed6","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"refusenextcount","value_label":"","hasOther":false,"value":"","path":"root/address/refusenextcount","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"recyclenextdate":{"name":"recyclenextdate","type":"text","id":"AF-Field-14665ddc-d4e1-4b88-b287-ef314ba29ad5","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"recyclenextdate","value_label":"","hasOther":false,"value":"","path":"root/address/recyclenextdate","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"recycle_freq":{"name":"recycle_freq","type":"text","id":"AF-Field-c324f54b-a6df-4a04-ad0f-5c423efee8f6","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"recycle_freq","value_label":"","hasOther":false,"value":"","path":"root/address/recycle_freq","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"recycle_day":{"name":"recycle_day","type":"text","id":"AF-Field-0ad44ac5-2be3-4f36-9702-cb220a4eba24","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"recycle_day","value_label":"","hasOther":false,"value":"","path":"root/address/recycle_day","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"recyclenextcount":{"name":"recyclenextcount","type":"text","id":"AF-Field-31d2fc27-7cf4-4e9a-869d-cabb9a115e5f","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"recyclenextcount","value_label":"","hasOther":false,"value":"","path":"root/address/recyclenextcount","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"gardennextdate":{"name":"gardennextdate","type":"text","id":"AF-Field-812b0b09-6493-4bf2-99f8-88605bd14b45","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"gardennextdate","value_label":"","hasOther":false,"value":"","path":"root/address/gardennextdate","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"garden_freq":{"name":"garden_freq","type":"text","id":"AF-Field-e14df718-1100-4c34-aa1b-309f325b99f9","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"garden_freq","value_label":"","hasOther":false,"value":"","path":"root/address/garden_freq","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"garden_day":{"name":"garden_day","type":"text","id":"AF-Field-228aa066-8c19-415e-9b75-b4b9a22cbd5a","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"garden_day","value_label":"","hasOther":false,"value":"","path":"root/address/garden_day","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""},"gardennextcount":{"name":"gardennextcount","type":"text","id":"AF-Field-66b544b2-718b-4155-af48-066227f581e3","value_changed":true,"section_id":"AF-Section-f31139fa-56d3-4bac-998e-d2b5bf9de219","label":"gardennextcount","value_label":"","hasOther":false,"value":"","path":"root/address/gardennextcount","valid":"","totals":"","suffix":"","prefix":"","summary":"","hidden":false,"_hidden":true,"isSummary":false,"staticMap":false,"isMandatory":true,"isRepeatable":false,"currencyPrefix":"","decimalPlaces":"","hash":""}}},"isPublished":true,"formName":"Bin Calendar Subform","processId":"AF-Process-503f9daf-4db9-4dd8-876a-6f2029f11196","tokens":{"port":"","host":"contact.lincoln.gov.uk","site_url":"https://contact.lincoln.gov.uk/AchieveForms/?mode=fill&consentMessage=yes&form_uri=sandbox-publish://AF-Process-503f9daf-4db9-4dd8-876a-6f2029f11196/AF-Stage-a1c0af0f-fec1-4419-80c0-0dd4e1d965c9/definition.json&process=1&process_uri=sandbox-processes://AF-Process-503f9daf-4db9-4dd8-876a-6f2029f11196&process_id=AF-Process-503f9daf-4db9-4dd8-876a-6f2029f11196","site_path":"/AchieveForms/","site_origin":"https://contact.lincoln.gov.uk","user_agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36","site_protocol":"https","session_id":"bat4bpe49e41sgp15kps52ueid","product":"Self","formLanguage":"en","isAuthenticated":false,"api_url":"https://contact.lincoln.gov.uk/apibroker/","transactionReference":"","transaction_status":"","published":true,"sectionLength":1,"formUri":"sandbox://AF-Form-feb5f76c-85b0-4a14-be5a-842175b5b318","publishUri":"sandbox-publish://AF-Process-503f9daf-4db9-4dd8-876a-6f2029f11196/AF-Stage-a1c0af0f-fec1-4419-80c0-0dd4e1d965c9/definition.json","formId":"AF-Form-feb5f76c-85b0-4a14-be5a-842175b5b318","topFormId":"AF-Form-feb5f76c-85b0-4a14-be5a-842175b5b318","parentFormId":"AF-Form-feb5f76c-85b0-4a14-be5a-842175b5b318","formName":"Bin Calendar","topFormName":"Bin Calendar","parentFormName":"Bin Calendar","formDescription":"","topFormDescription":"","parentFormDescription":"","case_ref":"FS-Case-545464324","stage_id":"AF-Stage-a1c0af0f-fec1-4419-80c0-0dd4e1d965c9","processId":"AF-Process-503f9daf-4db9-4dd8-876a-6f2029f11196","stage_name":"Stage 1","processName":"Bin Calendar V2","stageLength":1,"processDescription":"","processUri":"sandbox-processes://AF-Process-503f9daf-4db9-4dd8-876a-6f2029f11196","version":"13","csrf_token":"b368eb823f9635a731ca71287dec2e1f49f85784","reference":"FS545464324","process_prefix":"FS-Case-","ipAddress":"82.31.108.39","countryCode":"GB","countryName":"United Kingdom","regionName":"","cityName":"","zipCode":"","latitude":"","longitude":"","timeZone":"Europe/London"},"env_tokens":{"internalurl":"lincolncity-dash.achieveservice.com","status_xmas_market":"closed","status_garden_waste_NEW":"open","status_garden_waste_RENEW":"open","status_xmas_coach_bookings":"closed","garden_waste_yearstart":"2023","garden_waste_yeardisplay":"2023/24","garden_waste_yearend":"2024","garden_waste_closed":"closed","fee_garden_waste":"39","fee_garden_waste_delivery":"15","fee_garden_waste_extrasticker":"15","appraisal_targetdate":"30/09/2021","xmas_redirect":"http://www.lincoln-christmasmarket.co.uk/","dbr_respondbydate":"2022-12-05","Corporate_Parking_Year":"2023","dbr_year":"2021","parking_season_broadgate_annual_fee":"1020.00","parking_season_lawn_annual_fee":"1020","parking_season_lucytower_annual_fee":"1300","parking_season_tentercroft_annual_fee":"1300","parking_season_central_annual_fee":"1300.00","parking_season_broadgate_monthly_fee":"90.00","parking_season_lawn_monthly_fee":"90","parking_season_lucytower_monthly_fee":"115","parking_season_tentercroft_monthly_fee":"115","parking_season_central_monthly_fee":"115","parking_season_central_transponder_fee":"10","snn_housename":"1730","snn_app_fee":"5730","snn_plot_fee":"1440","dbr_closedate":"2023-01-06","status_xmas_parkandride":"closed","status_xmas_busking":"closed","status_xmas_businesspass":"closed"},"site":"aHR0cHM6Ly9jb250YWN0LmxpbmNvbG4uZ292LnVrL0FjaGlldmVGb3Jtcy8/bW9kZT1maWxsJmNvbnNlbnRNZXNzYWdlPXllcyZmb3JtX3VyaT1zYW5kYm94LXB1Ymxpc2g6Ly9BRi1Qcm9jZXNzLTUwM2Y5ZGFmLTRkYjktNGRkOC04NzZhLTZmMjAyOWYxMTE5Ni9BRi1TdGFnZS1hMWMwYWYwZi1mZWMxLTQ0MTktODBjMC0wZGQ0ZTFkOTY1YzkvZGVmaW5pdGlvbi5qc29uJnByb2Nlc3M9MSZwcm9jZXNzX3VyaT1zYW5kYm94LXByb2Nlc3NlczovL0FGLVByb2Nlc3MtNTAzZjlkYWYtNGRiOS00ZGQ4LTg3NmEtNmYyMDI5ZjExMTk2JnByb2Nlc3NfaWQ9QUYtUHJvY2Vzcy01MDNmOWRhZi00ZGI5LTRkZDgtODc2YS02ZjIwMjlmMTExOTY=","processName":"Bin Calendar V2","created":"2023-09-08 00:22:25","reference":"FS545464324","formUri":"sandbox-publish://AF-Process-503f9daf-4db9-4dd8-876a-6f2029f11196/AF-Stage-a1c0af0f-fec1-4419-80c0-0dd4e1d965c9/definition.json"}' \
  --compressed
```
-- edit by @5ila5 changed code snipped formatting

**Comment by 5ila5** [comment]
Source will be part of release 1.43.0 or is available now as master (git) version

**Comment by RoseOO** [comment]
Just tested and it is working great thank you!

One slight change suggestion, as I know it only gets the next collection so eventually it will overlap when the next weeks collection has passed. But in Lincoln if there is a recycling collection and refuse collection will happen on the same day. Would it be possible to artificially populate that next refuse collection?

As I've shown with the line:

![image](https://github.com/mampfes/hacs_waste_collection_schedule/assets/50427842/551bd2ff-2e61-4acc-ac02-f191b780c252)

**Comment by 5ila5** [comment]
implemented change in #1249

**Comment by kiwi3007** [comment]
Refuse is not collected at the same time as recycling where I live in Lincoln. Is it possible this option can be toggled? Or perhaps considered a different source? I've reverted the change on my local version, but this will get overwritten with updates.

**Comment by 5ila5** [comment]
Fixed in #2781.
This implementation uses the actual returned collection frequency to determine the future collections, so it should match for everyone.

This will be part of the next release or is available now as master (git) version 

install master/git: Developer tools -> actions -> select `update: Install update` -> click `Coose entity` and select WCS -> select `version` and type `master` in the input field -> run the action -> restart HA


<details><summary>Older versions of HACS:</summary>
<p>
install master with HACS version < 2.0: (3-dot menu -> redownload -> change version to master -> fully restart HA (this will increase update cycle by a lot)
</p>
</details>
